HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,343 @@
+  desc.add<std::vector<int> >("MinNstationsL2", std::vector<int>(1, 1));
+  desc.add<std::vector<int> >("MinNhitsL2", std::vector<int>(1, 0));
+  desc.add<std::vector<int> >("MinNchambersL2", std::vector<int>(1, 0));
+  descriptions.add("hltMuonL3orL2PreFilter", desc);
+}
+
+// ------------ method called to produce the data  ------------
+bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+
+  if (saveTags())
+    filterproduct.addCollectionTag(candTag_);
+
+  // get hold of trks
+  Handle<RecoChargedCandidateCollection> mucands;
+  iEvent.getByToken(candToken_, mucands);
+  Handle<TriggerFilterObjectWithRefs> previousLevelCands;
+  iEvent.getByToken(previousCandToken_, previousLevelCands);
+  Handle<RecoChargedCandidateCollection> L2mucands;
+  iEvent.getByToken(L2CandToken_, L2mucands);
+  vector<RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(TriggerMuon, vcands);
+  }
+
+  Handle<BeamSpot> recoBeamSpotHandle;
+  iEvent.getByToken(beamspotToken_, recoBeamSpotHandle);
+
+  // Number of objects passing the L3 Trigger or L2:
+  int n = 0;
+  bool firstIteration = true;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
+    RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "cand isNonnull " << L2cand.isNonnull();
+
+    //    //did this candidate triggered at previous stage.
+    //    if (matchPreviousCand_ && !triggerdByPreviousLevel(cand, vcands))
+    //      continue;
+    bool FailedL2 = false;
+    if (std::abs(L2cand->eta()) > max_Eta_L2_)
+      FailedL2 = true;
+
+    TrackRef L2tk = L2cand->track();
+    LogDebug("HLTMuonL3orL2PreFilter") << " Muon in loop, q*pt= " << L2tk->charge() * L2tk->pt() << " ("
+                                       << L2cand->charge() * L2cand->pt() << ") "
+                                       << ", eta= " << L2tk->eta() << " (" << L2cand->eta() << ") "
+                                       << ", hits= " << L2tk->numberOfValidHits() << ", d0= " << L2tk->d0()
+                                       << ", dz= " << L2tk->dz();
+
+    //normalizedChi2 cut
+    if (L2tk->normalizedChi2() > max_NormalizedChi2_L2_)
+      FailedL2 = true;
+
+    if (recoBeamSpotHandle.isValid()) {
+      const BeamSpot& beamSpot = *recoBeamSpotHandle;
+
+      //dz cut
+      if (std::abs((L2cand->vz() - beamSpot.z0()) -
+                   ((L2cand->vx() - beamSpot.x0()) * L2cand->px() + (L2cand->vy() - beamSpot.y0()) * L2cand->py()) /
+                       L2cand->pt() * L2cand->pz() / L2cand->pt()) > max_Dz_)
```

Comment:
```suggestion
      if (std::abs((L2cand->vz() - beamSpot.z0()) -
                   ((L2cand->vx() - beamSpot.x0()) * L2cand->px() + (L2cand->vy() - beamSpot.y0()) * L2cand->py()) /
                       L2cand->pt() * L2cand->pz() / L2cand->pt()) > max_Dz_)
```
The expression is fairly hard to read. Can you add a comment with an explanation of this cut, and try to make the code more readable?

Reply:
Yes, it is a cut on the impact parameter of the muon in the z direction. I will make it clearer.